### ISSUE TITLE
Modernize the graphical rendering system

### ISSUE THREAD

**Issue by Wohlstand** [Wohlstand/TheXTech#272]
For future needs, we need to modernize the graphical rendering system to extend its flexibility.

## What game does currently
In-game elements were grouped by logical parts:
- Background image
- Background-2 BGO
- Sizable blocks
- Background-1 BGO
- Background NPC
- Normal blocks
- Playable characters
- Foreground NPC
- Foreground-1 BGO
- Lava blocks (called "Foreground blocks")
- Foreground-2 BGO
- HUD

And there is no way to adjust some more exact Z-position of objects, because of a strictly-fixed grouping of objects

## How it can be improved
We could give all in-game elements the sort of Z-value which should help to organize elements more flexibly and more freely. However, if don't properly care about the performance, it may cause some sort of the performance slow downing ([noticed](https://github.com/Wohlstand/TheXTech/issues/271#issuecomment-1027031572) by @ds-sloth).

The modernized render mechanism will be required in the future for scripts and organized render of objects. Also, multi-layer background support will require this for the ability to draw some of its parts as foreground fragments, and there are some demos shown at the Moondust Engine with that: https://www.youtube.com/watch?v=g2tvgAK-eYM

At Moondust Engine, I have the next thing here: https://github.com/WohlSoft/Moondust-Project/blob/master/Engine/scenes/level/lvl_camera.cpp
- All visible objects gets a place at the array/queue of visible objects
- New-added elements will be appended into the list, and the list will be sorted (the fact of the object appears/disappears on the camera, must be registered at the object itself by a boolean flag, which should help us to avoid often rescanning of the queue for the object's presence)
- Elements that go off the visible area will be popped without breaking the queue
- When there are no elements had appear/disappeared, an array of visible elements stay unchanged

The algorithm may be improved to reduce any possible performance shocks and we should test the final thing on weak devices to see the result in action. For example, make a sort of a linked list that should allow inserting any new-added objects.

However, there are better ways of the thing improve can be suggested. We have two goals to reach:
- Extend the flexibility of on-screen elements ordering
- Avoid possible reduction of performance on weak devices

And, if there are needs for dynamic allocation/reallocation, we should use the custom allocator (similar to that, which I made at LunaScript to resolve the problem of dynamic memory allocation abuse).

**Comment by Wohlstand** [comment]
Do you mean the clubbing strange effect that happens when many items with similar Z values? At Moondust Engine I had another crap from that: sibling objects had to randomly blink and change their appearance every frame. So, I found the solution with adding to Z value the UID multiplied with 0.00000000001 (or, alternatively, a global counter of objects), which guarantees nobody have the same Z-value.

**Comment by ds-sloth** [comment]
> strange effect that happens when many items with similar Z values

No, but that is also an issue. I was thinking about the case where a semi-transparent object is drawn with a low (near) Z-value and then an opaque object is drawn with a high (far) Z-value, then the opaque object will not be drawn because the depth buffer is already set near in that area.

**Comment by Wohlstand** [comment]
What about the software case, the method I made at Moondust Engine ensures the render list doesn't change often and has the cheaper resorting at final. Instead of the array, we could try to use the sort of a linked list or red-black tree (use custom allocator!) to fill it quickly and don't resort it manually, but, the case could be benchmarked separately to see how much it affects the performance (will try with the well-known quicksort, and the red-black tree (at std::map, or craft something custom, but again, with a custom allocator))

**Comment by Wohlstand** [comment]
> Do you think that 8 bits is sufficient to represent Z-value or should we use 16 bits?

There are two keys needed: the Z-value, and in-array order index. Looks like the 16bit one is required, and have it work as a fixed-point number. So, the first byte will have the -128~+128 range for the integer part, and the second byte from 0...255 as decimal part for case where are scripts wish to specify a more precise order.

**Comment by ds-sloth** [comment]
> linked list
I think that as long as XRender maintains the render list (which I think is a good idea since we may have both hardware and software implementations) then we will need to completely re-fill the render list every frame. If we are refilling the list every frame, then we could use a custom allocator but there is no reason to since we will always deallocate the whole array at the same time. Using a simple array also has the advantage that instead of a `std::pair<uint16_t, DrawCall*>` pair, which could be 80 bits, we can use a single `uint32_t` where the upper bytes are the 16bit fixed-point number you describe and the lower bytes are the index in our array. Sorting `uint32_t` will be much cheaper than sorting `std::pair<uint16_t, DrawCall*>`.

If you want to avoid refilling the list every frame, then we will need to deeply rewrite the game to basically use a sprite system, which I do not think would be worth the difficulty.

> So, the first byte will have the -128~+128 range for the integer part, and the second byte from 0...255 as decimal part for cases where are scripts wish to specify a more precise order., or enforce order at the same group (of same integer Z).

That sounds good.

**Comment by ds-sloth** [comment]
Sounds good, and as long as this is happening inside of `XRender` it is an implementation detail we can change as we like. The main front-facing change would be to update the signature of the render functions to accept a `uint16_t` Z-index. Would you like to do this now?

We can set it so that the Z-index defaults to 0 and we enforce that for a single Z-index, draws will occur in order as current, so that the code can be updated gradually.

**Comment by Wohlstand** [comment]
> We can set it so that the Z-index defaults to 0 and we enforce that for a single Z-index, draws will occur in order as current, so that the code can be updated gradually.

Could add the argument into the tail of every draw call, and use the `0xFFFF` as a default value to mean, "draw now", or something also. We could use lambdas here to simplify the stuff internally.

P.S. I did some check and looks like the array filling and its post-sorting isn't so expensive:
[sort_benchmarks.cpp.zip](https://github.com/Wohlstand/TheXTech/files/7981532/sort_benchmarks.cpp.zip)

```
$ ./sort_benchmarks 
Sorting of array...
Sorted in 330172127 ns
Storing into red-black tree...
Filled in 540648473 ns
```

**Comment by ds-sloth** [comment]
Maybe 0 is good if we are signed, so that some can be drawn behind of legacy calls, some in front.

**Comment by Wohlstand** [comment]
I tried to use structures as data instead of integer:
[sort_benchmarks.cpp.zip](https://github.com/Wohlstand/TheXTech/files/7981576/sort_benchmarks.cpp.zip)

results are same:
```
$ ./sort_benchmarks 
Sorting of array of structs...
Sorted in 381227723 ns
Storing into red-black tree of structs...
Filled in 552927155 ns
```
> The benchmark is very interesting. Are the results the same for uint32_t? We may want to check again when we have begun implementation.

Gonna verify....

EDIT: Same
```
$ ./sort_benchmarks 
Sorting of array of structs...
Sorted in 386612264 ns
Storing into red-black tree of structs...
Filled in 558476195 ns
```
[sort_benchmarks.cpp.zip](https://github.com/Wohlstand/TheXTech/files/7981585/sort_benchmarks.cpp.zip)

**Comment by ds-sloth** [comment]
Cool, it is good to see that the results are consistent. Would you like me to code this up? I can probably have it done in two days.

**Comment by Wohlstand** [comment]
You may try some experiments yeah, would be very helpful :fox_face:

**Comment by suicvne** [comment]
This discussion reminds me of the work I've done on my vgl_renderer.c. While primitive, slightly messy, and definitely not yet optimized enough, it uses a linear list of a DrawCall struct to draw quads to the screen. 

So please, let me clarify what I'm seeing here:

* DrawCalls will be some sort of struct, not necessarily the one I used that is definitely *not* the best example. 
    * The 'key' for these DrawCalls will loosely represent a 'z index' that is multiplied by a small value to ensure uniqueness. The use of a uint16_t will allow for `65,535` theoretical max draw calls. Thoughts on that? Seems like a good maximum number for a 2D side-scroller to me.
* They will be filed into a std::vector, out of order but with their 'z index' taken into account
* Sorted either after each call or right before a final draw aka 'flush' by their keys using C++'s `std::sort`
* Drawn in the sorted order from back to front.  

The DrawCall struct system has potential to be platform agnostic if done properly, it would just be up to the renderer to consume these draw calls accordingly.

**Comment by suicvne** [comment]
So far I think it sounds great. But my question is:
> There could be two vectors -- one that directly owns all of the DrawCalls, and one that includes the keys for the DrawCalls. Only the second one would be sorted.

Do we need 2 Vectors to keep track of this? Why not use a `std::vector<std::pair<keytype_t, drawcall_t>>` and then sort by the key? 

Even if we're using the upper 16 bits and lower 16 bits of a 32-bit integer, wouldn't `std::sort`ing this vector by the key still get us the result we're looking for?

**Comment by ds-sloth** [comment]
> Ah well, couldn't doing that mess with the memory layout and introduce a bottleneck?

Yes, that's exactly what I'm thinking, we want to keep all of the keys in cache during the sort. Additionally, the sort would become more complex. Exchanging the position of two `uint32_t`s requires 3 instructions max. Suppose a `drawcall_t` is 16 bytes (might be even bigger). Then exchanging the position of two of these pairs will require many more instructions, maybe more than ten. Additionally, the comparison will probably become much more complicated than the single instruction comparison of two `uint32_t`s.

**Comment by ds-sloth** [comment]
I implemented something like this in 25982ae66b8510fcd0c790673ed02a1ac272c38c, d0d6d0f6149e94bf9f824a6100372462e0522e2d, 3e89a43744db57399612073253f3f57b4405f26b.

It has 32 primary user-specifiable draw planes (enumerated in 1868abf70bd3da352edb5799b1e3a7351162ad7d), with 7 offset planes per primary plane. Each primary plane can order 1824 draws per frame, and each offset plane can order 32 draws per frame. This fills the draw ordering capacity of the lowest-end targets, which use 16-bit hardware depth buffers.

Is this sufficient for the goals of this issue?
